server/embedded/src/org/labkey/embedded/LabKeyServer.java

```diff
@@ -0,0 +1,358 @@
+				var webAppLocation = "";
+
+				if (!webAppLocationPresent)
+				{
+					var destDirectory = new File("").getAbsolutePath() + "/server";
+					webAppLocation = destDirectory + "/labkeywebapp";
+					boolean extracted = new File(webAppLocation).exists();
+
+					if (!extracted)
+					{
+						try (JarFile jar = new JarFile("labkeyServer-21.1-SNAPSHOT.jar"))
+						{
+							boolean foundDistributionZip = false;
+							Enumeration<JarEntry> entries = jar.entries();
+							while (entries.hasMoreElements())
+							{
+								JarEntry entry = entries.nextElement();
+								String entryName = entry.getName();
+
+								if ("labkey/distribution.zip".equals(entryName))
+								{
+									foundDistributionZip = true;
+									try (var distInputStream = jar.getInputStream(entry))
+									{
+										LabKeyServer.extractZip(distInputStream, destDirectory);
+									}
+								}
+							}
+
+							if (!foundDistributionZip)
+							{
+								throw new ConfigException("Unable to find distribution zip required to run LabKey server.");
+							}
+						}
+						catch (IOException | ConfigException e)
+						{
+							throw new RuntimeException(e);
+						}
+					}
+				}
+				else
+				{
+					webAppLocation = props.getWebAppLocation();
+				}
+
+				// TODO : 8021 :fix context path - put app at root by default
+				StandardContext context = (StandardContext)tomcat.addWebapp("/labkey", webAppLocation);
+
+                // Push the JDBC connection for the primary DB into the context so that the LabKey webapp finds them
+				getDataSourceResources(props).forEach(contextResource -> context.getNamingResources().addResource(contextResource));
+				// Add the SMTP config
+				context.getNamingResources().addResource(getMailResource());
+
+				// And the master encryption key
+				context.addParameter("MasterEncryptionKey", props.getMasterEncryptionKey());
+
+				// Point at the special classloader with the hack for SLF4J
+				WebappLoader loader = new WebappLoader();
+				loader.setLoaderClass(LabKeySpringBootClassLoader.class.getName());
+				context.setLoader(loader);
+				context.setParentClassLoader(this.getClass().getClassLoader());
+
+				return super.getTomcatWebServer(tomcat);
+			}
+
+			private List<ContextResource> getDataSourceResources(ContextProperties props)
+			{
+				List<ContextResource> dataSourceResources = new ArrayList<>();
+				var numOfDataResources = props.getUrl().size();
+
+				for (int i = 0; i < numOfDataResources; i++)
+				{
+					ContextResource dataSourceResource = new ContextResource();
+					dataSourceResource.setName(props.getDataSourceName().get(i));
+					dataSourceResource.setAuth("Container");
+					dataSourceResource.setType(DataSource.class.getName());
+					dataSourceResource.setProperty("driverClassName", props.getDriverClassName().get(i));
+					dataSourceResource.setProperty("url", props.getUrl().get(i));
+					dataSourceResource.setProperty("password", props.getPassword().get(i));
```

Comment:
This will blow up with an ugly stack trace if you have more URLs than you have passwords, etc.